src/editors/editorsController.ts

```diff
@@ -229,117 +232,130 @@ export default class EditorsController {
       );
 
-      if (activeConnectionId !== connectionId) {
-        // Send metrics to Segment.
-        this._telemetryController.trackDocumentUpdated('treeview', false);
+      this._statusView.hideMessage();
 
-        vscode.window.showErrorMessage(
-          `Unable to save document: no longer connected to '${connectionName}'`
-        );
+      // Send metrics to Segment.
+      this._telemetryController.trackDocumentUpdated('treeview', true);
 
-        return resolve(false);
-      }
+      editor.document.save();
+      vscode.window.showInformationMessage(
+        `The document was saved successfully to '${namespace}'`
+      );
 
-      const dataservice = this._connectionController.getActiveDataService();
+      return true;
+    } catch (error) {
+      this._statusView.hideMessage();
 
-      if (dataservice === null) {
-        // Send metrics to Segment.
-        this._telemetryController.trackDocumentUpdated('treeview', false);
+      const errorMessage = `Unable to save document: ${error.message}`;
 
-        vscode.window.showErrorMessage(
-          `Unable to save document: no longer connected to '${connectionName}'`
-        );
+      // Send metrics to Segment.
+      this._telemetryController.trackDocumentUpdated('treeview', false);
 
-        return resolve(false);
-      }
+      vscode.window.showErrorMessage(errorMessage);
 
-      this._statusView.showMessage('Saving document...');
+      return false;
+    }
+  }
 
-      let newDocument: EJSON.SerializableTypes = {};
+  async saveDocumentToMongoDB(): Promise<boolean> {
+    const activeEditor = vscode.window.activeTextEditor;
 
-      try {
-        newDocument = EJSON.parse(activeEditor?.document.getText());
-      } catch (error) {
-        // Send metrics to Segment.
-        this._telemetryController.trackDocumentUpdated('treeview', false);
+    log.info('save document to MongoDB', activeEditor);
 
-        vscode.window.showErrorMessage(error.message);
+    if (!activeEditor) {
+      return true;
+    }
 
-        return resolve(false);
-      }
+    const uriParams = new URLSearchParams(activeEditor.document.uri.query);
+    const documentLocation =
+        uriParams.get(DOCUMENT_LOCATION_URI_IDENTIFIER) || '';
+    const namespace = uriParams.get(NAMESPACE_URI_IDENTIFIER) || '';
+    const connectionId = uriParams.get(CONNECTION_ID_URI_IDENTIFIER);
+    const documentIdReference =
+        uriParams.get(DOCUMENT_ID_URI_IDENTIFIER) || '';
+    const documentId = this._documentIdStore.get(documentIdReference);
 
-      dataservice.findOneAndReplace(
-        namespace,
-        {
-          _id: documentId
-        },
-        newDocument,
-        {
-          returnOriginal: false
-        },
-        (error) => {
-          this._statusView.hideMessage();
+    if (
+      documentLocation !== 'mongodb' ||
+        !namespace ||
+        !connectionId ||
+        !documentId
+    ) {
+      vscode.commands.executeCommand('workbench.action.files.save');
```

Comment:
The code from line `302` seems to contain a valuable piece of business logic that doesn't really stands out to someone that doesn't have a tons of context:

What is this condition for (is it cause the doc was never saved before? or some other special condition?), why do we save the file in this case and not in the other?

```
documentLocation !== 'mongodb' ||
        {	        !namespace ||
          _id: documentId	        !connectionId ||
        },	        !documentId
```

I think it would be helpful to break this method down with more functions to clarify a bit this behaviour or at least add some comment

Reply:
Looks like this checks to make sure the document being saved has all of the correct uri parameters inside of vscode. Maybe there's a way the uri can be edited when a document is open, so this check hopefully protects against any malformed uris.
I think checking for `documentId` to be true here might make a user unable to save a document with an `_id` of `false`. Not sure if that's a bug we want to fix. 🤔 
I'd like to keep this PR mostly just about the linting, not much on the logic changing side, so I'll just leave a comment for now if that's cool.

Reply:
json document can be mongodb document or just any other json document that we don't want to save to mongodb. we check if the uri have all required properties that we need in order to save it to mongodb we don't get inside this if and later in code call `replaceDocument` function. But if not all the data presented it means there is some trouble with this document. Maybe user tried to  malform uri or this is just not the mongodb document that we opened from the sidebar - in this case instead of saving it to mongodb we just handle this like a regular editor and ask user if they want to save it on disc as a file.

Reply:
@alenakhineika Thanks for the clarification! Yeah .. i think this as a comment would help a lot, this is quite deep to understand for someone who did not participated in the implementation.

I often find factoring this contrived concepts in dedicated methods helpful, for example with this:

```
_isMongodbDocument(document): boolean {
  const uriParams = new URLSearchParams(document.uri.query);
  const documentLocation = uriParams.get(DOCUMENT_LOCATION_URI_IDENTIFIER) || '';
  const namespace = uriParams.get(NAMESPACE_URI_IDENTIFIER) || '';
  const connectionId = uriParams.get(CONNECTION_ID_URI_IDENTIFIER);
  const documentIdReference =
      uriParams.get(DOCUMENT_ID_URI_IDENTIFIER) || '';
  const documentId = this._documentIdStore.get(documentIdReference);
  
  return documentLocation === 'mongodb' && namespace && connectionId && documentId;  
}
```

That `if` becomes:

``` js
if (!this._isMongodbDocument(activeEditor.document)) {
    vscode.commands.executeCommand('workbench.action.files.save'); // save to disk instead of mongodb
}
```

which both reduces the complexity of the main function, and also could help to put all those checks in their context.

Reply:
I like it a lot! will refactor it!